enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,283 @@
+	// claims about the image.
+	SignedIdentity Identity `json:"signedIdentity,omitempty"`
+}
+
+type Identity struct {
+	// Defaults to "matchRepository"
+	IdentityMatchPolicy IdentityMatchPolicy `json:"identityMatchPolicy,…"`
+	DockerRepository    string              `json:"dockerRepository,omitempty"`
+}
+
+type IdentityMatchPolicy string
+
+const (
+	IdentityMatchPolicyMatchRepository = "matchRepository"
+	IdentityMatchPolicyExactRepository = "exactRepository"
+)
+
+type ImagePolicyStatus struct {
+	// PolicyJSON contains the whole policy applied to the namespace
+	// which got written to disk. This includes cluster-wide policies
+	// from the `openshift-config` namespace as well.
+	PolicyJSON string `json:"policyJSON,omitempty"`
+
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Update container runtime config controller to watch ImagePolicy
+
+MCO container runtime config controller ImagePolicy syncHandler updates the signature verification configurations:
+- retrieves all the ImagePolicy instances on the cluster.
+- validates conflicting data from ImagePolicy instances.
+- adds following configurations to a single machine config
+  - merges the instances for each namespace into a single [containers-policy.json](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md) in a predefined /path/to/policies/\<NAMESPACE\>.json.  
+  If no policy exists in a namespace, use the current default of:
+
+    ```yaml
+    { "default": [{ "type": "insecureAcceptAnything" }] }
+    ```
+
+    If a single policy exists, use the default:
+    
+    ```yaml
+    { "default": [{"type": "reject"}] }
+    ```
+
+  - adds [/etc/containers/registries.d/*.yaml](https://github.com/containers/image/blob/main/docs/containers-registries.d.5.md) to allow matching sigstore signatures, for example:
+
+    ```yaml
+    docker:
+      my-registry/image:
+        use-sigstore-attachments: true
+    ```
+
+    Alternatively, we can enable it node wide:
+
+    ```yaml
+    default-docker:
+      use-sigstore-attachments: true
+    ```
+- merges cluster scoped with namespaced ImagePolicy as a json string and update `status` of each ImagePolicy.
+Once an ImagePolicy object gets created/updated, container runtime config controller will call the syncHandler.
+Once an ImagePolicy object gets deleted, container runtime config controller will delete the machine config and run the syncHandler
+
+#### Notes
+
+ImagePolicy is defined as namespaced CRD. `namespace: openshift-config` is used for cluster policy.
+
+### Risks and Mitigations
+
+Risk: The cluster scoped and namespace policies overwrite
+
+Mitigation: MCO container runtime config controller makes sure namespaced policies for an image cannot override global definitions when merging cluster openshift-config policies and namespaced policies into status field of ImagePolicy instances.   
```

Comment:
Could you elaborate on this point?

Reply:
The syncHandler in container runtime config controller will merge all the imagePolicy from opeshift-config namespace and other namespaces, it should error out if the sigsore policy for the same image scope has been configured under global(`openshift-config`) namespace and non global namspace.

Reply:
> ... it should error out if the sigsore policy for the same image scope has been configured under global(1openshift-config1) namespace and non global namspace.

[The policy spec includes][1]:

> The policy requirements are represented as a JSON array of individual requirement objects.  For an image to be accepted, *all* of the requirements must be satisfied simultaneously.

So maybe we can just append entries when creating the policy being used to pull images for workloads in that namespace?  Or is this limited by the CRI, and today there is no way to say "I'm pulling this for `$NAMESPACE`, so use its policies"?  If we are limited to a single unified policy, doesn't this API allow for a denial-of-service conflict where a workload admin trying to configure namespace A sets a strict policy that breaks the ability to pull images into namespace B workloads?

[1]: https://github.com/containers/image/blame/8ad019310ff2ed8f99dea75b18998b415955bf8e/docs/containers-policy.json.5.md#L117-L119